### PR TITLE
ringdown update

### PR DESCRIPTION
**Pull request by JulianWesterweck**
Adds ability to use arbitrary polarisations for the ringdown modes.
Adds option to use spheroidal harmonics instead of spherical harmonics.
Includes -m modes.

### CODE REVIEW

File: pycbc/conversions.py [hunk 1]
```diff
@@ -972,10 +963,18 @@ def freq_from_final_mass_spin(final_mass, final_spin, l=2, m=2, nmodes=1):
         ``[input shape x] nmodes``, where ``input shape`` is the broadcasted
         shape of the inputs.
     """
-    return get_lm_f0tau(final_mass, final_spin, l, m, nmodes)[0]
+    if n is not None:
+        nmodes = n + 1
+    out = get_lm_f0tau(final_mass, final_spin, l, m, nmodes, which='f0')
+    if n is not None or nmodes == 1:
+        if n is None:
```
Comment (micamu) [inline pycbc/conversions.py:970]: Could we set the default `n=0` (changing the if statements accordingly)? Then we would not need this additional `if` statement

File: pycbc/waveform/ringdown.py [hunk 2]
```diff
@@ -395,34 +504,148 @@ def apply_taper(tau, amp, phi, delta_t, l=2, m=2, inclination=None):
...
+    hplus : TimeSeries
+        The plus polarization.
+    hcross : TimeSeries
+        The cross polarization.
+    """
+    # create the times
+    tlen = int(t_final/delta_t) + 1
+    times = numpy.linspace(0, t_final, num=tlen)
+    # evaluate the harmonics
+    if inclination is None:
```
[the remark this answers is not in the document]
Comment (micamu) [reply pycbc/waveform/ringdown.py:611]: I guess the same would apply to the azimuthal angle

File: pycbc/waveform/ringdown.py [hunk 3]
```diff
@@ -485,7 +711,18 @@ def multimode_base(input_params, domain, freq_tau_approximant=False):
...
-    amps, phis = lm_amps_phases(**input_params)
+    amps, phis, dbetas, dphis = lm_amps_phases(**input_params)
+    pols, polnms = lm_arbitrary_harmonics(**input_params)
+    # get harmonics argument
+    try:
+        harmonics = input_params['harmonics']
+    except KeyError:
+        harmonics = 'spherical'
+    # add inclination and azimuthal if they aren't provided
+    if 'inclination' not in input_params:
```
Comment (micamu) [inline pycbc/waveform/ringdown.py:722]: This is where the warning or error suggested in my previous comment should go, to make sure the user is aware if what happens if they forget to provide an inclination and azimuth angle.

File: pycbc/waveform/ringdown.py [hunk 4]
```diff
@@ -501,20 +738,36 @@ def multimode_base(input_params, domain, freq_tau_approximant=False):
...
         for lmn in freqs:
-            hplus, hcross = td_damped_sinusoid(freqs[lmn], taus[lmn],
-                            amps[lmn], phis[lmn], outplus.delta_t,
-                            outplus.sample_times[-1], int(lmn[0]), int(lmn[1]),
-                            input_params['inclination'])
+            #hplus, hcross = td_damped_sinusoid(freqs[lmn], taus[lmn],
+            #                amps[lmn], phis[lmn], outplus.delta_t,
+            #                outplus.sample_times[-1], int(lmn[0]), int(lmn[1]),
+            #                input_params['inclination'], pols[lmn])
+            if harmonics == 'spheroidal':
```
Comment (micamu) [inline pycbc/waveform/ringdown.py:750]: Wouldn't the spherical harmonics also want a final_spin parameter?

File: pycbc/conversions.py [hunk 5]
```diff
@@ -912,10 +892,13 @@ def get_lm_f0tau_allmodes(mass, spin, modes):
...
     modes : list of str
-        The modes to get. Each string in the list should be formatted 'lmN',
-        where l (m) is the l (m) index of the harmonic and N is the number of
-        overtones to generate (note, N is not the index of the overtone). For
-        example, '221' will generate the 0th overtone of the l = m = 2 mode.
+        The modes to get. Each string in the list should be formatted
+        'lmN[nm]', where l (m) is the l (m) index of the harmonic and N is the
+        number of overtones to generate (note, N is not the index of the
+        overtone). If a mode has the string "nm" added to the end of it (for
+        negative m), the -m mode will be used. For example, '221' will generate
```
Comment (micamu) [inline pycbc/conversions.py:899]: Do I understand correctly that, if I ask for `221nm` I will only get the (2,-2,0) mode? So if I want both positive and negative modes, I need to request both `221` and `221nm`?
If that isn't elsewhere in the code already (I am having a hard time following all the changes), could we also have an option that if one requests `221{x}` then one would get *both* the +m and the -m mode (with `x` any name of your choice)?

**Review by micamu — COMMENTED** [review]
It might be worth going through the changes together, I am having a hard time following the motivation for or understanding the implications of all the changes

File: pycbc/conversions.py (hunk 1 above)
Comment (cdcapano) [reply pycbc/conversions.py:971]: Good point... I'll change it to default to n=0, and get rid of the extra if statement.

File: pycbc/conversions.py [hunk 6]
```diff
@@ -1000,7 +1000,15 @@ def tau_from_final_mass_spin(final_mass, final_spin, l=2, m=2, nmodes=1):
...
-    return get_lm_f0tau(final_mass, final_spin, l, m, nmodes)[1]
+    if n is not None:
+        nmodes = n + 1
+    out = get_lm_f0tau(final_mass, final_spin, l, m, nmodes, which='tau')
+    if (n is not None or nmodes == 1) and (
+        isinstance(out, numpy.ndarray) and out.ndim > 0):
+        if n is None:
+            n = 0
+        out = out[..., n]
+    return out
```
Comment (cdcapano) [inline pycbc/conversions.py:1011]: Make same fix (default n=0) here as above.

File: pycbc/conversions.py [hunk 7]
```diff
@@ -1009,16 +1017,23 @@ def tau_from_final_mass_spin(final_mass, final_spin, l=2, m=2, nmodes=1):
...
 
 _berti_mass_constants = {
     (2, 2): (1.5251, -1.1568, 0.1292),
+    (2, -2): (0.2938, 0.0782, 1.3546),
+    (2, 1): (0.6, -0.2339, 0.4175),
+    (2, 0): (0.4437, 0.0739, 0.3350),
     (3, 3): (1.8956, -1.3043, 0.1818),
     (4, 4): (2.3, -1.5056, 0.2244),
-    (2, 1): (0.6, -0.2339, 0.4175)
     }
```
Comment (cdcapano) [inline pycbc/conversions.py:1037]: More to do: comment out the -m and 2,0 constants for now, since we're not certain how well they work.

File: pycbc/waveform/ringdown.py [hunk 8]
```diff
@@ -25,10 +25,14 @@
...
-import re, numpy, lal
+import re, numpy, lal, lalsimulation
+try:
+    import pykerr
+except ImportError:
+    pykerr = None
 from pycbc.types import TimeSeries, FrequencySeries, float64, complex128, zeros
 from pycbc.waveform.waveform import get_obj_attrs
-from pycbc.conversions import get_lm_f0tau_allmodes
+from pycbc.conversions import (get_lm_f0tau_allmodes, MINUS_M_SUFFIX)
```
Comment (cdcapano) [inline pycbc/waveform/ringdown.py:35]: To do: remove MINUS_M_SUFFIX import and usage below.

File: pycbc/waveform/ringdown.py [hunk 9]
```diff
@@ -82,7 +86,10 @@ def format_lmns(lmns):
     # Catch case of lmns given as float (as int injection values are cast
     # to float by pycbc_create_injections), cast to int, then string
     if isinstance(lmns, float):
-        lmns = str(int(lmns))
+        if lmns < 0:
+            lmns = str(abs(int(lmns))) + MINUS_M_SUFFIX
+        else:
+            lmns = str(int(lmns))
```
Comment (cdcapano) [inline pycbc/waveform/ringdown.py:92]: Remove this, set back to what it was originally.

File: pycbc/waveform/ringdown.py [hunk 10]
```diff
@@ -102,60 +109,96 @@ def format_lmns(lmns):
         # to a string
         # lmn = lmn.strip(" b'")
         # Try to convert to int and then str, to ensure the right format
+        minusm = lmn.endswith(MINUS_M_SUFFIX)
+        if minusm:
+            lmn = lmn.replace(MINUS_M_SUFFIX, '')
```
Comment (cdcapano) [inline pycbc/waveform/ringdown.py:114]: Remove this.

File: pycbc/waveform/ringdown.py [hunk 11]
```diff
@@ -102,60 +109,96 @@ def format_lmns(lmns):
...
         lmn = str(int(lmn))
         if len(lmn) != 3:
             raise ValueError('Format of parameter lmns not recognized. See '
                              'approximant documentation for more info.')
         elif int(lmn[2]) == 0:
             raise ValueError('Number of overtones (nmodes) must be greater '
                              'than zero in lmn={}.'.format(lmn))
+        if minusm:
+            lmn = lmn + MINUS_M_SUFFIX
```
Comment (cdcapano) [inline pycbc/waveform/ringdown.py:123]: Remove minus m suffix.

File: pycbc/waveform/ringdown.py [hunk 12]
```diff
@@ -102,60 +109,96 @@ def format_lmns(lmns):
...
+    """
+    lm, nmodes = lmn[0:2], int(lmn[2])
+    minusm = lmn.endswith(MINUS_M_SUFFIX)
+    overtones = []
+    for n in range(nmodes):
+        mode = lm + '{}'.format(n)
+        if minusm:
+            mode += MINUS_M_SUFFIX
+        overtones.append(mode)
+    return overtones
```
Comment (cdcapano) [inline pycbc/waveform/ringdown.py:139]: Remove the minus m suffix thing, but keep this function.

File: pycbc/waveform/ringdown.py [hunk 13]
```diff
@@ -537,7 +790,8 @@ def multimode_base(input_params, domain, freq_tau_approximant=False):
                             input_params['f_lower'],
                             input_params['f_final'],
                             l=int(lmn[0]), m=int(lmn[1]),
-                            inclination=input_params['inclination'])
+                            inclination=input_params['inclination'],
+                            pol=pols[lmn])
             outplus.data += hplus.data
             outcross.data += hcross.data
     else:
```
Comment (cdcapano) [inline pycbc/waveform/ringdown.py:797]: Other to do: Add documentation for new parameters (`ref_amp`, `dbeta`, `dphi`, etc.) to the various `get_` functions that are outward facing, but refer them to `td_damped_sinusoid` for the math.

File: pycbc/waveform/ringdown.py [hunk 14]
```diff
@@ -501,20 +738,36 @@ def multimode_base(input_params, domain, freq_tau_approximant=False):
...
+                pol=pols[lmn], polnm=polnms[lmn])
             if input_params['taper'] and outplus.delta_t < taus[lmn]:
                 taper_hp, taper_hc = apply_taper(taus[lmn], amps[lmn],
                                      phis[lmn], outplus.delta_t, int(lmn[0]),
-                                     int(lmn[1]), input_params['inclination'])
+                                     int(lmn[1]), input_params['inclination'],
+                                     pols[lmn])
                 t0 = -int(outplus.start_time * outplus.sample_rate)
                 outplus[t0-len(taper_hp):t0].data += taper_hp
                 outplus[t0:].data += hplus
```
Comment (cdcapano) [inline pycbc/waveform/ringdown.py:773]: Update the frequency domain version so that it works with the new arguments.

File: pycbc/waveform/ringdown.py (hunk 4 above)
Comment (cdcapano) [reply pycbc/waveform/ringdown.py:750]: No, don't need final spin for the spherical harmonics.

File: pycbc/waveform/ringdown.py [hunk 15]
```diff
@@ -323,60 +373,119 @@ def td_output_vector(freqs, damping_times, taper=False,
...
+    raise ValueError("FIX ME")
     taper_times = -numpy.arange(0, int(tau/delta_t))[::-1] * delta_t
-    if inclination is not None:
+    if pol is not None:
+        Y_plus = numpy.cos(pol)
+        Y_cross = numpy.sin(pol)
+    elif inclination is not None:
         Y_plus, Y_cross = spher_harms(l, m, inclination)
     else:
         Y_plus, Y_cross = 1, 1
```
Comment (cdcapano) [inline pycbc/waveform/ringdown.py:491]: To do: remove taper.

File: pycbc/waveform/ringdown.py [hunk 16]
```diff
@@ -395,34 +504,148 @@ def apply_taper(tau, amp, phi, delta_t, l=2, m=2, inclination=None):
...
+        The cross polarization.
+    """
+    # create the times
+    tlen = int(t_final/delta_t) + 1
+    times = numpy.linspace(0, t_final, num=tlen)
+    # evaluate the harmonics
+    if inclination is None:
+        inclination = 0.
+    if azimuthal is None:
+        azimuthal = 0.
```
Comment (cdcapano) [inline pycbc/waveform/ringdown.py:614]: To do: remove if statements on inclination and azimuthal since they both default to 0.

File: pycbc/conversions.py (hunk 1 above)
Comment (cdcapano) [reply pycbc/conversions.py]: I simplified this. Now all functions in conversions.py just take an `n` that can be an int or an array, like the other indices. This works with the ringdown module, since the ringdown module is only calling `get_lm_f0tau_allmodes` with the modes dictionary.

File: pycbc/waveform/ringdown.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (cdcapano) [reply pycbc/waveform/ringdown.py]: Removed.

File: pycbc/conversions.py (hunk 5 above)
Comment (cdcapano) [reply pycbc/conversions.py]: Removed.

File: pycbc/conversions.py (hunk 7 above)
Comment (cdcapano) [reply pycbc/conversions.py:970]: Removed.

File: pycbc/waveform/ringdown.py (hunk 8 above)
Comment (cdcapano) [reply pycbc/waveform/ringdown.py]: Done

File: pycbc/waveform/ringdown.py (hunk 9 above)
Comment (cdcapano) [reply pycbc/waveform/ringdown.py]: Done

File: pycbc/waveform/ringdown.py (hunk 10 above)
Comment (cdcapano) [reply pycbc/waveform/ringdown.py]: Done

File: pycbc/waveform/ringdown.py (hunk 11 above)
Comment (cdcapano) [reply pycbc/waveform/ringdown.py]: Done

File: pycbc/waveform/ringdown.py (hunk 12 above)
Comment (cdcapano) [reply pycbc/waveform/ringdown.py:130]: Done

File: pycbc/waveform/ringdown.py (hunk 13 above)
Comment (cdcapano) [reply pycbc/waveform/ringdown.py]: Done

File: pycbc/waveform/ringdown.py (hunk 14 above)
Comment (cdcapano) [reply pycbc/waveform/ringdown.py]: Done, although figuring out how to get non-circular polarization (via the dphi and dbeta parameters) was more work than I wanted to do right now, so the frequency domain version doesn't support that. This is the same functionality as before, with the added ability to use spheroidal or arbitrary harmonics.

File: pycbc/waveform/ringdown.py (hunk 15 above)
Comment (cdcapano) [reply pycbc/waveform/ringdown.py]: I left the taper capabilities in, but fixed them. Tapering had been broken for quite some time.

File: pycbc/waveform/ringdown.py (hunk 16 above)
Comment (cdcapano) [reply pycbc/waveform/ringdown.py]: Done

**Comment by cdcapano** [comment]
@micamu I think I've addressed all the issues. I fixed the frequency domain version to work with the updates (although it doesn't support non-circular polarization), fixed tapering, added documentation, and simplified the functions in conversions.py. The notebook I used to test the functions can be downloaded from [here](https://www.atlas.aei.uni-hannover.de/~work-cdcapano/scratch/prs/ringdown_updates/).

File: pycbc/waveform/ringdown.py [hunk 17]
```diff
@@ -882,32 +930,60 @@ def get_fd_from_final_mass_spin(template=None, **kwargs):
...
-        lm pair (maximum n=8).
-        Example: lmns = ['223','331'] are the modes 220, 221, 222, and 330
-    amp220 : float
-        Amplitude of the fundamental 220 mode.  Always required, even if 220
-        mode has not been selected. Note that if distance is given,
-        this parameter will have a completely different order of magnitude.
-        See table II in https://arxiv.org/abs/1107.0854 for an estimate.
+        lm pair, not the overtone number; maximum n=8. Example:
+        lmns = ['223','331'] are the modes 220, 221, 222, and 330
+    ref_amp : str, optional
```
Comment (micamu) [inline pycbc/waveform/ringdown.py:945]: Why is this needed?

Comment (micamu) [reply pycbc/waveform/ringdown.py:945]: Oh, I see in the function `lm_amps_phases` that it defaults to amp220 if no ref_amp is given, so I guess it is ok

**Review by micamu — APPROVED** [review]
